rust/protocol/src/sealed_sender.rs

```diff
@@ -879,8 +1019,214 @@ mod sealed_sender_v2 {
+        // Derive an ephemeral key pair from those random bytes.
+        let ephemeral_keys = DerivedKeys::calculate(&m);
+        let ephemeral_public_key = ephemeral_keys.e.public_key;
+
+        // Encrypt the ephemeral key pair.
+        let sender_c_0: [u8; MESSAGE_KEY_LEN] = apply_agreement_xor(
+            &ephemeral_keys.e,
+            recipient_identity.public_key(),
+            Direction::Sending,
+            &m,
+        )?
+        .as_ref()
+        .try_into()
+        .expect("fixed-size slice");
+        // Compute an authentication tag for the encrypted key pair.
+        let sender_at_0 = compute_authentication_tag(
+            &sender_identity,
+            recipient_identity.identity_key(),
+            Direction::Sending,
+            &ephemeral_public_key,
+            &sender_c_0,
+        )?;
+
+        // The message recipient calculates the original random bytes and authenticates the result.
+        let recv_m = apply_agreement_xor(
+            &recipient_identity.into(),
+            &ephemeral_public_key,
+            Direction::Receiving,
+            &sender_c_0,
+        )?;
+        assert_eq!(recv_m.as_ref(), m.as_ref());
+
+        let recv_at_0 = compute_authentication_tag(
+            &recipient_identity,
+            sender_identity.identity_key(),
+            Direction::Receiving,
+            &ephemeral_public_key,
+            &sender_c_0,
+        )?;
+        assert_eq!(recv_at_0.as_ref(), sender_at_0.as_ref());
+
+        Ok(())
+    }
 }
 
+/// This method implements a single-key multi-recipient [KEM] as defined in Manuel Barbosa's
+/// ["Randomness Reuse: Extensions and Improvements"], a.k.a. Sealed Sender v2.
+///
+/// [KEM]: https://en.wikipedia.org/wiki/Key_encapsulation
+/// ["Randomness Reuse: Extensions and Improvements"]: https://haslab.uminho.pt/mbb/files/reuse.pdf
+///
+/// # Contrast with Sealed Sender v1
+/// The KEM scheme implemented by this method uses the "Generic Construction" in `4.1` of [that
+/// paper]["Randomness Reuse: Extensions and Improvements"], instantiated with [ElGamal
+/// encryption]. This technique enables reusing a single sequence of random bytes across multiple
+/// messages with the same content, which reduces computation time for clients sending the same
+/// message to multiple recipients (without compromising the message security).
+///
+/// There are a few additional design tradeoffs this method makes vs [Sealed Sender v1]
+/// which may make it comparatively unwieldy for certain scenarios:
+/// 1. it requires a [SessionRecord] to exist already for the recipient, i.e. that a Double Ratchet
+///    message chain has previously been established in the [SessionStore] via
+///    [crate::process_prekey_bundle] after an initial [crate::PreKeySignalMessage] is recieved.
+/// 2. it ferries a lot of additional information in its encoding which makes the resulting message
+///    bulkier than the message produced by this method.
```

Comment:
"this method" is left over from v1, but in addition I'll add a second sentence of my own: "For sending, this will generally still be more compact than sending the same message N times, but on the receiver side the message is slightly larger."

Reply:
Yes, that's much better!